packages/daimo-api/src/contract/ClogMatcher.ts

```diff
@@ -0,0 +1,287 @@
+          .leftJoin("index.daimo_fast_cctp as start", (join) =>
+            join
+              .onRef("start.chain_id", "=", "ff.from_chain_id")
+              .onRef("start.handoff_addr", "=", "ff.handoff_addr")
+              .onRef("start.nonce", "=", "ff.nonce")
+              .on("start.log_name", "=", "Start")
+          )
+          .select([
+            "start.block_hash as block_hash",
+            "start.block_num as block_num",
+            "start.tx_hash as tx_hash",
+            "start.tx_idx as tx_idx",
+            "start.log_idx as log_idx",
+            "start.from_token as from_token",
+            "start.from_addr as from_addr",
+            "start.to_addr as to_addr",
+            "start.to_amount as to_amount",
+            "start.from_chain_id as from_chain_id",
+          ])
+          .where("ff.chain_id", "=", "" + chainId)
+          .where((e) => e.between("ff.block_num", "" + from, "" + to))
+          .where("ff.tx_hash", "in", txList)
+          .where("ff.log_name", "=", "FastFinish")
+          .execute()
+    );
+
+    for (const row of fastCctpRes) {
+      const startLog = {
+        blockHash: bytesToHex(assertNotNull(row.block_hash), { size: 32 }),
+        blockNumber: BigInt(assertNotNull(row.block_num)),
+        transactionHash: bytesToHex(assertNotNull(row.tx_hash), { size: 32 }),
+        transactionIndex: Number(row.tx_idx),
+        logIndex: Number(row.log_idx),
+        address: getAddress(
+          bytesToHex(assertNotNull(row.from_token), { size: 20 })
+        ),
+        from: getAddress(
+          bytesToHex(assertNotNull(row.from_addr), { size: 20 })
+        ),
+        to: getAddress(bytesToHex(assertNotNull(row.to_addr), { size: 20 })),
+        value: BigInt(assertNotNull(row.to_amount)),
+      };
+
+      const crossChainToken = this.tokenReg.getToken(
+        startLog.address,
+        Number(row.from_chain_id) // cross-chain (from chain)
+      );
+      if (crossChainToken == null) continue;
+
+      const crossChainTransfer: ForeignTokenTransfer = {
+        ...startLog,
+        foreignToken: crossChainToken,
+      };
+      this.txHashToTransfers.set(crossChainTransfer.transactionHash, [
+        ...(this.txHashToTransfers.get(crossChainTransfer.transactionHash) ||
+          []),
+        crossChainTransfer,
+      ]);
+    }
+
+    const elapsedMs = (Date.now() - startTime) | 0;
+    console.log(
+      `[SWAP] loaded ${fastCctpRes.length} FastCCTP cross-chain events in ${elapsedMs}ms`
+    );
+  }
+
+  /**
+   * For a multi-log outbound transfer (e.g. SwapClog, cross-chain transfer),
+   * attach the corresponding transfer info.
+   */
+  getCorrespondingTransfer(
+    from: Address,
+    to: Address,
+    transactionHash: string,
+    isInbound?: boolean
+  ) {
+    const transfers = this.txHashToTransfers.get(transactionHash);
+    if (!transfers || transfers.length === 0) return null;
+
+    // First check whether the transfer is cross-chain.
```

Comment:
to your point about whether its fastfinish, i don't know how else a non-"this chain" transfer would get in the mix -- maybe we should add log_name to the log type though, wdyt